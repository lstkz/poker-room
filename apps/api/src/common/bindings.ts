import * as R from 'remeda';
import {
  BaseBinding,
  CreateEventBindingOptions,
  CreateRpcBindingOptions,
} from '../lib';

const bindings = [
  require('../contracts/user/getMe'),
  require('../contracts/user/login'),
  require('../contracts/user/register'),
  require('../contracts/table/createTable'),
  require('../contracts/table/getAllTables'),
  require('../contracts/table/joinTable'),
  require('../contracts/table/leaveTable'),
  require('../contracts/table/getTableById'),
  require('../contracts/table/createTable'),
  require('../contracts/game/getCurrentGame'),
  require('../contracts/game/makeMove'),
  require('../contracts/example/createFoo'),
  require('../contracts/example/getAll'),
];

export function getBindings(type: 'rpc'): CreateRpcBindingOptions[];
export function getBindings(type: 'event'): CreateEventBindingOptions<any>[];
export function getBindings(
  type: 'rpc' | 'event'
): CreateRpcBindingOptions[] | CreateEventBindingOptions<any>[] {
  return R.pipe(
    bindings,
    R.flatMap(obj => Object.values(obj) as BaseBinding<string, any>[]),
    R.filter(x => x.isBinding && x.type === type),
    R.map(x => x.options)
  );
}
