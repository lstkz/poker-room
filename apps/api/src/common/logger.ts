import { LOG_LEVEL } from '../config';

export const logger = {
  error(...args: any) {
    console.error('ERROR:', ...args);
  },
  debug(...args: any) {
    if (LOG_LEVEL === 'debug') {
      console.log('DEBUG:', ...args);
    }
  },
  info(...args: any) {
    console.log('INFO:', ...args);
  },
};
