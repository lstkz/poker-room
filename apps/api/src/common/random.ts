import crypto from 'mz/crypto';

export async function randomInt() {
  return (await crypto.randomBytes(4)).readUInt32BE(0);
}
