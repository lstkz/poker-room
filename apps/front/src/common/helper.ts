import { AuthData } from 'shared';
import { GlobalActions } from 'src/features/global/interface';
import * as Rx from 'src/rx';
import { setAccessToken } from 'src/services/Storage';
import { RouterActions } from 'typeless-router';

interface HandleAuthOptions {
  Actions: {
    setLoading(isLoading: boolean): any;
    setError(error: string): any;
  };
  auth(): Rx.Observable<AuthData>;
}

export function handleAuth(options: HandleAuthOptions) {
  const { Actions, auth } = options;
  return Rx.concatObs(
    Rx.of(Actions.setLoading(true)),
    Rx.of(Actions.setError('')),
    auth().pipe(
      Rx.mergeMap(authData => {
        setAccessToken(authData.accessToken);
        return [GlobalActions.loggedIn(authData.user), RouterActions.push('/')];
      }),
      Rx.catchError(e => {
        const error = e.response?.error ?? e.message;
        return Rx.of(Actions.setError(error));
      })
    ),
    Rx.of(Actions.setLoading(false))
  );
}
