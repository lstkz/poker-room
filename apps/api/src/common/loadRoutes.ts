import * as R from 'remeda';
import { Router } from 'express';
import fs from 'fs';
import Path from 'path';
import { wrapExpress } from './wrapExpress';
import { BadRequestError, UnauthorizedError } from './errors';
import { logger } from './logger';
import { Handler } from '../types';
import { RpcBinding } from '../lib';

const bindings = R.flattenDeep(
  fs.readdirSync(Path.join(__dirname, '../contracts')).map(dir => {
    return fs
      .readdirSync(Path.join(__dirname, '../contracts', dir))
      .map(fileName => {
        const contractModule: Record<
          string,
          RpcBinding
        > = require(`../contracts/${dir}/${fileName}`);
        return Object.values(contractModule)
          .filter(x => x.isBinding && x.type === 'rpc')
          .map(x => x.options);
      });
  })
);

export default function loadRoutes(router: Router) {
  bindings.forEach(options => {
    const actions: Handler[] = [
      (req, res, next) => {
        if (!req.headers.authorization) {
          return next();
        }
        return next(new Error('Authentication not implemented'));
      },
      (req, res, next) => {
        if (options.public) {
          next();
          return;
        }
        if (!req.user) {
          next(new UnauthorizedError('Bearer token required'));
          return;
        }
        next();
      },
    ];
    if (process.env.NODE_ENV !== 'test') {
      logger.info(
        `${options.public ? '[Public]' : '[Auth]'} ${options.signature}`
      );
    }
    actions.push((req, res, next) => {
      if (req.body == null) {
        req.body = {};
      } else if (typeof req.body !== 'object') {
        next(new BadRequestError('Body must be an object'));
        return;
      }
      const params = options.handler.getParams();
      if (options.injectUser) {
        params.shift();
      }
      const diff = R.difference(Object.keys(req.body), params);
      if (diff.length) {
        next(
          new BadRequestError(
            'Following params are not allowed: ' + diff.join(', ')
          )
        );
        return;
      }
      const values = params.map(x => req.body[x]);
      if (options.injectUser) {
        values.unshift(undefined); // todo
      }
      Promise.resolve(options.handler(...values))
        .then(ret => {
          res.json(ret);
        })
        .catch(next);
    });
    router.post('/' + options.signature, wrapExpress(actions) as any);
  });
}
