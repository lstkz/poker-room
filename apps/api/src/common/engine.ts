import * as R from 'remeda';
import { Card, MoveType } from 'shared';
import { GameModel } from '../collections/Game';
import { InvalidMoveError, UnreachableCaseError } from './errors';
import { randomInt } from './random';

export class CardRandomizer {
  private cards: Card[] = [];

  constructor(usedCards: Card[] = []) {
    const allCards = [2, 3, 4, 5, 6, 7, 8, 9, 'T', 'J', 'Q', 'K', 'A'] as const;
    const allColors = ['c', 'd', 'h', 's'] as const;
    const usedSet = new Set<string>(
      usedCards.map(card => card.card + card.color)
    );
    allCards.forEach(card => {
      allColors.forEach(color => {
        const key = card + color;
        if (!usedSet.has(key)) {
          this.cards.push({ card, color });
        }
      });
    });
  }

  getLeftCards() {
    return this.cards;
  }

  async randomNextCard(): Promise<Card> {
    if (!this.cards.length) {
      throw new Error('No cards left');
    }
    const idx = (await randomInt()) % this.cards.length;
    const ret = this.cards.splice(idx, 1);
    if (!ret.length) {
      return this.randomNextCard();
    }
    return ret[0];
  }
}

export function getActPlayer(game: GameModel) {
  const phase = R.last(game.phases)!;
  const lastMove = R.last(phase.moves);
  const getPlayer = (idx: number) => game.players[idx % game.players.length];

  if (lastMove) {
    const lastPlayerIdx = game.players.findIndex(x =>
      x.userId.equals(lastMove.userId)
    );
    const skipPlayers = new Set(
      phase.moves
        .filter(x => x.moveType === 'all-in' || x.moveType === 'fold')
        .map(x => x.userId.toHexString())
    );
    for (let i = 1; i < game.players.length; i++) {
      const nextPlayerIdx = (lastPlayerIdx + i) % game.players.length;
      const player = game.players[nextPlayerIdx];
      if (!skipPlayers.has(player.userId.toHexString())) {
        return player;
      }
    }
    throw new Error('Cannot get next act player');
  } else {
    const dealerPlayerIdx = game.players.findIndex(
      x => x.seat === game.dealerPosition
    );
    const posDiff =
      phase.type === 'pre-flop'
        ? 3 // utg-1
        : 1; // sb;
    return getPlayer(dealerPlayerIdx + posDiff);
  }
}

export function getBB(stakesOrGame: number | GameModel) {
  const stakes =
    typeof stakesOrGame === 'number' ? stakesOrGame : stakesOrGame.stakes;
  return stakes / 100;
}

export function getMinRaiseAmount(game: GameModel) {
  const bb = getBB(game);
  if (!game.currentBets.length) {
    return bb;
  }
  if (game.currentBets.length === 1) {
    // bb: 0.5$
    // raise: 5$
    // min re-raise: 9.5$
    return game.currentBets[0] * 2 - bb;
  }
  const last = R.last(game.currentBets)!;
  const diff = last - game.currentBets[game.currentBets.length - 2];
  return last + diff;
}

export function validateMove(
  values: {
    moveType: MoveType;
    raiseAmount?: number | undefined;
  },
  game: GameModel
) {
  const player = getActPlayer(game);
  const currentBet = R.last(game.currentBets);
  switch (values.moveType) {
    case 'fold':
      // always can fold
      break;
    case 'all-in':
      // always can all-in
      break;
    case 'check':
      {
        if (currentBet) {
          if (game.betMap[player.userId.toHexString()] !== currentBet) {
            throw new InvalidMoveError("Can't check if another player bet");
          }
        }
      }
      break;
    case 'call':
      {
        if (!currentBet) {
          throw new InvalidMoveError("Can't call if no bet");
        }
        if (game.betMap[player.userId.toHexString()] === currentBet) {
          throw new InvalidMoveError("Can't call if nothing to call");
        }
        if (player.money < currentBet) {
          throw new InvalidMoveError('No enough money for call');
        }
      }
      break;
    case 'raise':
      {
        if (!values.raiseAmount) {
          throw new InvalidMoveError('Amount required for raise');
        }
        const minRaise = getMinRaiseAmount(game);
        if (values.raiseAmount < minRaise) {
          throw new InvalidMoveError(
            `Raise amount min raise: $${minRaise.toFixed(2)}`
          );
        }
        if (values.raiseAmount >= player.money) {
          throw new InvalidMoveError('Not enough money for raise');
        }
      }
      break;
    default:
      new UnreachableCaseError(values.moveType);
  }
}

export function processMove(
  values: {
    moveType: MoveType;
    raiseAmount?: number | undefined;
  },
  game: GameModel
) {
  const phase = R.last(game.phases)!;
  const player = getActPlayer(game);
  const currentBet = R.last(game.currentBets)!;
  switch (values.moveType) {
    case 'fold':
      {
        phase.moves.push({
          amount: 0,
          moveType: 'fold',
          userId: player.userId,
        });
      }
      break;
    case 'all-in':
      {
        phase.moves.push({
          amount: player.money,
          moveType: 'all-in',
          userId: player.userId,
        });
        game.betMap[player.userId.toHexString()] = player.money;
        game.currentBets.push(player.money);
      }
      break;
    case 'check':
      {
        phase.moves.push({
          amount: 0,
          moveType: 'check',
          userId: player.userId,
        });
      }
      break;
    case 'call':
      {
        phase.moves.push({
          amount: currentBet,
          moveType: 'call',
          userId: player.userId,
        });
        game.betMap[player.userId.toHexString()] = currentBet;
      }
      break;
    case 'raise':
      {
        const raiseAmount = values.raiseAmount!;
        phase.moves.push({
          amount: raiseAmount,
          moveType: 'raise',
          userId: player.userId,
        });
        game.betMap[player.userId.toHexString()] = raiseAmount;
        game.currentBets.push(raiseAmount);
      }
      break;
    default:
      new UnreachableCaseError(values.moveType);
  }
}

function getCurrentBet(game: GameModel) {
  if (!game.currentBets.length) {
    return getBB(game);
  }
  return R.last(game.currentBets);
}

export function getCardRandomizer(game: GameModel) {
  const used: Card[] = [];
  game.players.forEach(player => {
    used.push(...player.hand);
  });
  game.phases.forEach(phase => {
    used.push(...phase.cards);
  });
  return new CardRandomizer(used);
}

export async function processNextPhase(game: GameModel) {
  const currentBet = getCurrentBet(game);
  const playingPlayers = R.pipe(
    R.last(game.phases)!.moves,
    R.filter(x => x.moveType !== 'all-in' && x.moveType !== 'fold'),
    R.map(x => x.userId.toHexString())
  );
  if (!playingPlayers.every(playerId => game.betMap[playerId] === currentBet)) {
    return;
  }
  const cr = getCardRandomizer(game);
  // preflop -> flop
  if (game.phases.length === 1) {
    game.phases.push({
      moves: [],
      type: 'flop',
      cards: await Promise.all([
        cr.randomNextCard(),
        cr.randomNextCard(),
        cr.randomNextCard(),
      ]),
    });
    let sum = 0;
    const playerMap = R.indexBy(game.players, x => x.userId);
    Object.keys(game.betMap).forEach(userId => {
      const bet = game.betMap[userId];
      sum += bet;
      playerMap[userId].money -= bet;
    });
    game.pot += sum;
    game.betMap = {};
  }
}
