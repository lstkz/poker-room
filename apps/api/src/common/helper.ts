import crypto from 'mz/crypto';
import { randomInt } from './random';

export function renameId<T extends { _id: any }>(
  obj: T
): Omit<T, '_id'> & { id: string } {
  const ret: any = { ...obj };
  ret.id = ret._id.toString();
  delete ret._id;
  return ret;
}

export async function generateSalt() {
  const bytes = await crypto.randomBytes(256);
  return bytes.toString('hex');
}

export async function hashPassword(password: string, salt: string) {
  const buffer = await crypto.pbkdf2(
    password,
    salt,
    process.env.NODE_ENV === 'production' ? 100000 : 100,
    512,
    'sha512'
  );
  return buffer.toString('hex');
}

export async function randomString(len: number) {
  const charSet =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let randomString = '';
  for (let i = 0; i < len; i++) {
    let randomPoz = (await randomInt()) % charSet.length;
    randomString += charSet[randomPoz];
  }
  return randomString;
}

export function safeAssign<T>(obj: T, values: Partial<T>) {
  return Object.assign(obj, values);
}

export function safeKeys<T>(obj: T): Array<keyof T> {
  return Object.keys(obj) as any;
}

export async function randomItem<T>(items: T[]) {
  const idx = (await randomInt()) % items.length;
  return items[idx];
}
