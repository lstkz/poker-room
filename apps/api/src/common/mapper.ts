import * as R from 'remeda';
import { Table } from 'shared';
import { TableModel } from '../collections/Table';
import { UserModel } from '../collections/User';
import { renameId } from './helper';

export function mapTable(
  table: TableModel,
  userMap: Record<string, UserModel>
): Table {
  return {
    ...renameId(table),
    players: table.players
      .map(player => {
        const id = player.userId.toHexString();
        return {
          user: {
            id,
            username: userMap[id].username,
          },
          ...R.pick(player, ['money', 'seat']),
        };
      })
      .sort((a, b) => a.seat - b.seat),
  };
}
