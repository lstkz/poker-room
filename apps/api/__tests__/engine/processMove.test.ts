import { processNextPhase } from '../../src/common/engine';
import { chainMove, getPreflopGame, removePlayerHand } from './engine-helper';

describe('processMove', () => {
  it('fold', () => {
    const game = getPreflopGame();
    chainMove(game).fold();
    expect(removePlayerHand(game)).toMatchSnapshot();
  });

  it('fold -> fold', () => {
    const game = getPreflopGame();
    chainMove(game).fold().fold();
    expect(removePlayerHand(game)).toMatchSnapshot();
  });

  it('fold -> fold -> call', () => {
    const game = getPreflopGame();
    chainMove(game).fold().fold().call();
    expect(removePlayerHand(game)).toMatchSnapshot();
  });

  it('fold -> fold -> call -> check', () => {
    const game = getPreflopGame();
    chainMove(game).fold().fold().call().check();
    expect(removePlayerHand(game)).toMatchSnapshot();
  });

  it('raise -> fold -> call -> fold', () => {
    const game = getPreflopGame();
    chainMove(game).raise(2).fold().call().fold();
    expect(removePlayerHand(game)).toMatchSnapshot();
  });

  it('raise -> reraise -> fold -> fold -> call', () => {
    const game = getPreflopGame();
    chainMove(game).raise(2).raise(5).fold().fold().call();
    expect(removePlayerHand(game)).toMatchSnapshot();
  });

  it('no min raise on flop', async () => {
    const game = getPreflopGame();
    chainMove(game).raise(4).call().call().call();
    await processNextPhase(game);
    expect(() => chainMove(game).raise(0.5)).not.toThrow();
  });
});
