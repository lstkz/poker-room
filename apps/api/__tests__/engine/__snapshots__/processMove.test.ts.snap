// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processMove fold -> fold -> call -> check 1`] = `
Object {
  "_id": "000000010000000000000000",
  "betMap": Object {
    "000000000000000000000002": 0.5,
    "000000000000000000000004": 0.5,
  },
  "currentBets": Array [
    0.5,
  ],
  "dealerPosition": 1,
  "isPlaying": true,
  "phases": Array [
    Object {
      "moves": Array [
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000006",
        },
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000001",
        },
        Object {
          "amount": 0.5,
          "moveType": "call",
          "userId": "000000000000000000000002",
        },
        Object {
          "amount": 0,
          "moveType": "check",
          "userId": "000000000000000000000004",
        },
      ],
      "type": "pre-flop",
    },
  ],
  "players": Array [
    Object {
      "money": 50,
      "seat": 1,
      "userId": "000000000000000000000001",
    },
    Object {
      "money": 50,
      "seat": 2,
      "userId": "000000000000000000000002",
    },
    Object {
      "money": 50,
      "seat": 4,
      "userId": "000000000000000000000004",
    },
    Object {
      "money": 50,
      "seat": 6,
      "userId": "000000000000000000000006",
    },
  ],
  "pot": 0,
  "stakes": 50,
  "tableId": null,
}
`;

exports[`processMove fold -> fold -> call 1`] = `
Object {
  "_id": "000000010000000000000000",
  "betMap": Object {
    "000000000000000000000002": 0.5,
    "000000000000000000000004": 0.5,
  },
  "currentBets": Array [
    0.5,
  ],
  "dealerPosition": 1,
  "isPlaying": true,
  "phases": Array [
    Object {
      "moves": Array [
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000006",
        },
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000001",
        },
        Object {
          "amount": 0.5,
          "moveType": "call",
          "userId": "000000000000000000000002",
        },
      ],
      "type": "pre-flop",
    },
  ],
  "players": Array [
    Object {
      "money": 50,
      "seat": 1,
      "userId": "000000000000000000000001",
    },
    Object {
      "money": 50,
      "seat": 2,
      "userId": "000000000000000000000002",
    },
    Object {
      "money": 50,
      "seat": 4,
      "userId": "000000000000000000000004",
    },
    Object {
      "money": 50,
      "seat": 6,
      "userId": "000000000000000000000006",
    },
  ],
  "pot": 0,
  "stakes": 50,
  "tableId": null,
}
`;

exports[`processMove fold -> fold 1`] = `
Object {
  "_id": "000000010000000000000000",
  "betMap": Object {
    "000000000000000000000002": 0.25,
    "000000000000000000000004": 0.5,
  },
  "currentBets": Array [
    0.5,
  ],
  "dealerPosition": 1,
  "isPlaying": true,
  "phases": Array [
    Object {
      "moves": Array [
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000006",
        },
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000001",
        },
      ],
      "type": "pre-flop",
    },
  ],
  "players": Array [
    Object {
      "money": 50,
      "seat": 1,
      "userId": "000000000000000000000001",
    },
    Object {
      "money": 50,
      "seat": 2,
      "userId": "000000000000000000000002",
    },
    Object {
      "money": 50,
      "seat": 4,
      "userId": "000000000000000000000004",
    },
    Object {
      "money": 50,
      "seat": 6,
      "userId": "000000000000000000000006",
    },
  ],
  "pot": 0,
  "stakes": 50,
  "tableId": null,
}
`;

exports[`processMove fold 1`] = `
Object {
  "_id": "000000010000000000000000",
  "betMap": Object {
    "000000000000000000000002": 0.25,
    "000000000000000000000004": 0.5,
  },
  "currentBets": Array [
    0.5,
  ],
  "dealerPosition": 1,
  "isPlaying": true,
  "phases": Array [
    Object {
      "moves": Array [
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000006",
        },
      ],
      "type": "pre-flop",
    },
  ],
  "players": Array [
    Object {
      "money": 50,
      "seat": 1,
      "userId": "000000000000000000000001",
    },
    Object {
      "money": 50,
      "seat": 2,
      "userId": "000000000000000000000002",
    },
    Object {
      "money": 50,
      "seat": 4,
      "userId": "000000000000000000000004",
    },
    Object {
      "money": 50,
      "seat": 6,
      "userId": "000000000000000000000006",
    },
  ],
  "pot": 0,
  "stakes": 50,
  "tableId": null,
}
`;

exports[`processMove raise -> fold -> call -> fold 1`] = `
Object {
  "_id": "000000010000000000000000",
  "betMap": Object {
    "000000000000000000000002": 2,
    "000000000000000000000004": 0.5,
    "000000000000000000000006": 2,
  },
  "currentBets": Array [
    0.5,
    2,
  ],
  "dealerPosition": 1,
  "isPlaying": true,
  "phases": Array [
    Object {
      "moves": Array [
        Object {
          "amount": 2,
          "moveType": "raise",
          "userId": "000000000000000000000006",
        },
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000001",
        },
        Object {
          "amount": 2,
          "moveType": "call",
          "userId": "000000000000000000000002",
        },
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000004",
        },
      ],
      "type": "pre-flop",
    },
  ],
  "players": Array [
    Object {
      "money": 50,
      "seat": 1,
      "userId": "000000000000000000000001",
    },
    Object {
      "money": 50,
      "seat": 2,
      "userId": "000000000000000000000002",
    },
    Object {
      "money": 50,
      "seat": 4,
      "userId": "000000000000000000000004",
    },
    Object {
      "money": 50,
      "seat": 6,
      "userId": "000000000000000000000006",
    },
  ],
  "pot": 0,
  "stakes": 50,
  "tableId": null,
}
`;

exports[`processMove raise -> reraise -> fold -> fold -> call 1`] = `
Object {
  "_id": "000000010000000000000000",
  "betMap": Object {
    "000000000000000000000001": 5,
    "000000000000000000000002": 0.25,
    "000000000000000000000004": 0.5,
    "000000000000000000000006": 5,
  },
  "currentBets": Array [
    0.5,
    2,
    5,
  ],
  "dealerPosition": 1,
  "isPlaying": true,
  "phases": Array [
    Object {
      "moves": Array [
        Object {
          "amount": 2,
          "moveType": "raise",
          "userId": "000000000000000000000006",
        },
        Object {
          "amount": 5,
          "moveType": "raise",
          "userId": "000000000000000000000001",
        },
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000002",
        },
        Object {
          "amount": 0,
          "moveType": "fold",
          "userId": "000000000000000000000004",
        },
        Object {
          "amount": 5,
          "moveType": "call",
          "userId": "000000000000000000000006",
        },
      ],
      "type": "pre-flop",
    },
  ],
  "players": Array [
    Object {
      "money": 50,
      "seat": 1,
      "userId": "000000000000000000000001",
    },
    Object {
      "money": 50,
      "seat": 2,
      "userId": "000000000000000000000002",
    },
    Object {
      "money": 50,
      "seat": 4,
      "userId": "000000000000000000000004",
    },
    Object {
      "money": 50,
      "seat": 6,
      "userId": "000000000000000000000006",
    },
  ],
  "pot": 0,
  "stakes": 50,
  "tableId": null,
}
`;
